File: php/src/FloatFmt.php
<?php

declare(strict_types=1);

namespace Apstndb\SpanValue;

final class FloatFmt
{
    private static function isNegativeZero(float $v): bool
    {
        static $negZeroBytes = null;
        if ($negZeroBytes === null) {
            $negZeroBytes = pack('d', -0.0);
        }
        return pack('d', $v) === $negZeroBytes;
    }

    public static function narrowFloat32(float $v): float
    {
        $packed = pack('g', $v);
        $unpacked = unpack('g', $packed);
        return $unpacked[1];
    }

    private static function packFloat32(float $v): float
    {
        return self::narrowFloat32($v);
    }

    private static function roundTrips(string $s, float $original, int $bits): bool
    {
        if (!is_numeric($s)) {
            return false;
        }
        $parsed = (float) $s;
        if (is_nan($original)) {
            return is_nan($parsed);
        }
        if (is_infinite($original)) {
            return is_infinite($parsed) && ($parsed > 0) === ($original > 0);
        }
        if ($bits === 32) {
            return self::packFloat32($parsed) === self::packFloat32($original);
        }
        return $parsed === $original;
    }

    private static function fmtExponent(int $exp): string
    {
        if ($exp >= 0) {
            return sprintf('e+%02d', $exp);
        }
        $ae = abs($exp);
        if ($ae < 10) {
            return sprintf('e-%02d', $ae);
        }
        return 'e' . $exp;
    }

    private static function peToGoG(string $es): string
    {
        if (!preg_match('/^(\d)(?:\.(\d+))?e([+-])(\d+)$/', $es, $m)) {
            throw new \ValueError("unexpected e-format: {$es}");
        }
        $d1 = $m[1];
        $rest = $m[2] ?? '';
        $exp = (int) ($m[3] . $m[4]);
        $sig = $d1 . rtrim($rest, '0');
        if ($sig === '') {
            $sig = '0';
        }
        $ndigits = strlen($sig);

        if ($exp >= -4 && $exp < 6) {
            $decPos = 1 + $exp;
            if ($decPos <= 0) {
                $s = '0.' . str_repeat('0', -$decPos) . $sig;
            } elseif ($decPos >= $ndigits) {
                $s = $sig . str_repeat('0', $decPos - $ndigits);
            } else {
                $s = substr($sig, 0, $decPos) . '.' . substr($sig, $decPos);
            }
            if (str_contains($s, '.')) {
                $s = rtrim(rtrim($s, '0'), '.');
            }
            return $s;
        }

        if ($ndigits === 1) {
            $body = $sig;
        } else {
            $body = $sig[0] . '.' . substr($sig, 1);
        }
        return $body . self::fmtExponent($exp);
    }

    public static function formatGoG(float $v, int $bits = 64): string
    {
        if ($bits === 32) {
            $v = self::narrowFloat32($v);
        }

        if (is_nan($v)) {
            return 'NaN';
        }
        if (is_infinite($v)) {
            return $v < 0 ? '-Inf' : '+Inf';
        }
        if (self::isNegativeZero($v)) {
            return '-0';
        }

        $negative = $v < 0;
        $av = $negative ? -$v : $v;
        $maxP = $bits === 64 ? 16 : 8;

        $best = null;
        $target = $bits === 64 ? $v : self::narrowFloat32($v);
        for ($p = 0; $p <= $maxP; $p++) {
            $es = sprintf("%.{$p}e", $av);
            $g = self::peToGoG($es);
            $candidate = ($negative ? '-' : '') . $g;
            if (self::roundTrips($candidate, $target, $bits)) {
                if ($best === null || strlen($candidate) < strlen($best)) {
                    $best = $candidate;
                }
            }
        }

        if ($best === null) {
            $best = ($negative ? '-' : '') . (string) $av;
        }
        return $best;
    }

    private static function floatTrunc(float $v): float
    {
        return $v >= 0 ? floor($v) : ceil($v);
    }

    public static function formatSpannerCliFloat(float $v, int $bits = 64): string
    {
        if ($bits === 32) {
            $v = self::narrowFloat32($v);
        }
        if (is_nan($v)) {
            return 'NaN';
        }
        if (is_infinite($v)) {
            return $v < 0 ? '-Inf' : '+Inf';
        }
        if (self::isNegativeZero($v)) {
            return '-0';
        }
        if ($v === self::floatTrunc($v)) {
            return sprintf('%.0f', $v);
        }
        return sprintf('%.6f', $v);
    }

    public static function float64ToLiteral(float $v, LiteralQuoteConfig $quoteCfg): string
    {
        if (is_nan($v)) {
            return Quote::sqlCastQuoted('nan', 'FLOAT64', $quoteCfg);
        }
        if (is_infinite($v)) {
            return Quote::sqlCastQuoted($v < 0 ? '-inf' : 'inf', 'FLOAT64', $quoteCfg);
        }
        $s = self::formatGoG($v, 64);
        if (!strpbrk($s, '.eE')) {
            $s .= '.0';
        }
        return $s;
    }

    public static function float32ToLiteral(float $v, LiteralQuoteConfig $quoteCfg): string
    {
        $fv = self::narrowFloat32($v);
        if (is_nan($fv)) {
            return Quote::sqlCastQuoted('nan', 'FLOAT32', $quoteCfg);
        }
        if (is_infinite($fv)) {
            return Quote::sqlCastQuoted($fv < 0 ? '-inf' : 'inf', 'FLOAT32', $quoteCfg);
        }
        return 'CAST(' . self::formatGoG($fv, 32) . ' AS FLOAT32)';
    }
}

File: ruby/lib/spanvalue/float_fmt.rb
# frozen_string_literal: true

require 'fiddle'

module Spanvalue
  module FloatFmt
    E_RE = /\A(\d)(?:\.(\d+))?e([+-])(\d+)\z/i

    module_function

    def libc_parsers
      @libc_parsers ||= begin
        libc = Fiddle.dlopen(nil)
        {
          strtod: Fiddle::Function.new(
            libc['strtod'],
            [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_DOUBLE
          ),
          strtof: Fiddle::Function.new(
            libc['strtof'],
            [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_FLOAT
          )
        }
      end
    end

    def parse_float_string(s, bits)
      parsers = libc_parsers
      buf = Fiddle::Pointer[s + "\0"]
      endp = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
      if bits == 32
        parsers[:strtof].call(buf, endp).to_f
      else
        parsers[:strtod].call(buf, endp)
      end
    end

    def narrow_float32(v)
      [v].pack('e').unpack1('e')
    end

    def pack_float32(v)
      narrow_float32(v)
    end

    def round_trips?(s, original, bits)
      parsed = parse_float_string(s, bits)
      return original.nan? if parsed.nan?
      if parsed.infinite?
        return original.infinite? && (parsed <=> 0) == (original <=> 0)
      end

      if bits == 32
        return pack_float32(parsed) == pack_float32(original)
      end

      parsed == original
    rescue ArgumentError, TypeError
      false
    end

    def fmt_exponent(exp)
      if exp >= 0
        format('e+%02d', exp)
      elsif exp.abs < 10
        format('e-%02d', exp.abs)
      else
        "e#{exp}"
      end
    end

    def pe_to_go_g(es)
      m = E_RE.match(es)
      raise ArgumentError, "unexpected e-format: #{es.inspect}" unless m

      d1 = m[1]
      rest = m[2] || ''
      exp = Integer("#{m[3]}#{m[4]}")
      sig = (d1 + rest).sub(/0+\z/, '')
      sig = '0' if sig.empty?
      ndigits = sig.length

      if exp >= -4 && exp < 6
        dec_pos = 1 + exp
        s = if dec_pos <= 0
              "0.#{'0' * -dec_pos}#{sig}"
            elsif dec_pos >= ndigits
              "#{sig}#{'0' * (dec_pos - ndigits)}"
            else
              "#{sig[0...dec_pos]}.#{sig[dec_pos..]}"
            end
        if s.include?('.')
          s = s.sub(/0+\z/, '').sub(/\.\z/, '')
        end
        return s
      end

      body = ndigits == 1 ? sig : "#{sig[0]}.#{sig[1..]}"
      "#{body}#{fmt_exponent(exp)}"
    end

    def format_go_g(v, bits = 64)
      v = narrow_float32(v) if bits == 32

      return 'NaN' if v.nan?
      return '-Inf' if v.infinite? && v.negative?
      return '+Inf' if v.infinite?
      if v.zero? && (1.0 / v).infinite? && (1.0 / v).negative?
        return '-0'
      end

      negative = v.negative?
      av = v.abs
      max_p = bits == 64 ? 16 : 8
      target = bits == 64 ? v : narrow_float32(v)
      best = nil

      (0..max_p).each do |p|
        es = format("%.*e", p, av)
        g = pe_to_go_g(es)
        candidate = "#{negative ? '-' : ''}#{g}"
        next unless round_trips?(candidate, target, bits)

        best = candidate if best.nil? || candidate.length < best.length
      end

      best || "#{negative ? '-' : ''}#{av.inspect}"
    end

    def format_spanner_cli_float(v, bits = 64)
      v = narrow_float32(v) if bits == 32
      return 'NaN' if v.nan?
      return '-Inf' if v.infinite? && v.negative?
      return '+Inf' if v.infinite?

      if v == v.truncate
        format('%.0f', v)
      else
        format('%.6f', v)
      end
    end

    def float64_to_literal(v, quote_cfg)
      if v.nan?
        return Quote.sql_cast_quoted('nan', 'FLOAT64', quote_cfg)
      end
      if v.infinite?
        payload = v.negative? ? '-inf' : 'inf'
        return Quote.sql_cast_quoted(payload, 'FLOAT64', quote_cfg)
      end

      s = format_go_g(v, 64)
      s += '.0' unless s.match?(/[.eE]/)
      s
    end

    def float32_to_literal(v, quote_cfg)
      fv = narrow_float32(v)
      if fv.nan?
        return Quote.sql_cast_quoted('nan', 'FLOAT32', quote_cfg)
      end
      if fv.infinite?
        payload = fv.negative? ? '-inf' : 'inf'
        return Quote.sql_cast_quoted(payload, 'FLOAT32', quote_cfg)
      end

      "CAST(#{format_go_g(fv, 32)} AS FLOAT32)"
    end
  end
end
